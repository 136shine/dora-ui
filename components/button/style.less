.dora-btn {
  line-height: 2;
  cursor: pointer;
  outline: none;

  &__small {
    font-size: 12px;
    padding: 0 15px;
    border-radius: 2px;
  }

  &__medium {
    font-size: 14px;
    padding: 0 20px;
    border-radius: 3px;
  }

  &__large {
    font-size: 16px;
    padding: 0 30px;
    border-radius: 4px;
  }

  &__primary {
    background-color: blue;
  }

  &__warning {
    background-color: orange;
  }

  &__ghost {
    background-color: gray;
  }
}
