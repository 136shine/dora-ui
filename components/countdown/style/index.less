.dora-countdown {
  color: #fff;
  text-align: center;

  &-time {
    background-color: #bc3822;
    display: inline-block;
    width: 26px;
    height: 26px;
    line-height: 26px;
    font-size: 18px;
    border-radius: 2px;
  }

  &-unit {
    font-size: 12px;
    margin: 0 5px;
    height: 26px;
    vertical-align: top;
    line-height: 26px;
  }
}
