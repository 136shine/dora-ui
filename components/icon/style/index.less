@import '../../style/vars.less';

.dora-icon {
  overflow: hidden;
  vertical-align: -0.15em;
  fill: currentColor;

  &-sm {
    width: @icon-size-sm;
    height: @icon-size-sm;
  }

  &-md {
    width: @icon-size-md;
    height: @icon-size-md;
  }

  &-lg {
    width: @icon-size-lg;
    height: @icon-size-lg;
  }

  &-spinning {
    animation: spin 2s linear infinite;
  }

  @keyframes spin {
    to {
      transform: rotate(360deg);
    }
  }
}
