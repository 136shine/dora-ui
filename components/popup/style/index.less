@import '../../style/vars.less';

@popupPrefix: dora-popup;

.@{popupPrefix} {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  z-index: @popup-zindex;
  pointer-events: none; // amazing! 为none时可以产生点透的效果
  visibility: visible;
  opacity: 1;

  &__hide {
    // 为了离场动画效果 不能直接设置为display:none
    // opacity:0 visibility:hidden 约等于 display:none （文档流除外）
    opacity: 0;
    transition: all 0.2s;
    visibility: hidden; // magic! 也可以实现点透 包括子元素
  }

  .@{popupPrefix}-mask {
    position: absolute;
    width: 100%;
    height: 100%;
  }

  .@{popupPrefix}-mask {
    display: none; // mask 默认隐藏
    overflow: hidden;
    background-color: @fill-mask;
    pointer-events: auto;

    // fix some android webview opacity render bug
    &::before {
      content: '.';
      display: block;
      width: 1px;
      height: 1px;
      background-color: rgba(0, 0, 0, 0.1);
      margin-left: -10px;
    }
  }

  &__mask {
    pointer-events: auto;
    .@{popupPrefix}-mask {
      display: block; // 根类具有mask属性时展示mask
    }
  }

  .@{popupPrefix}-content {
    background-color: @toast-fill;
    color: @color-text-base-inverse;
    position: absolute;
    width: 100%;
    box-sizing: border-box;
    pointer-events: auto;
    overflow: scroll;
    -webkit-overflow-scrolling: touch; /* ios5+ */
  }

  &__left,
  &__right,
  &__center {
    .@{popupPrefix}-content {
      width: auto;
      max-width: 100%;
    }
  }

  &__left,
  &__right {
    .@{popupPrefix}-content {
      height: 100%;
    }
  }

  &__center {
    .@{popupPrefix}-content {
      border-radius: @radius-sm;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
  }

  &__right {
    .@{popupPrefix}-content {
      left: 100%;
      transform: translate(-100%, 0);
    }
  }

  &__bottom {
    .@{popupPrefix}-content {
      border-radius: @radius-md @radius-md 0 0;
      top: 100%;
      transform: translate(0, -100%);
    }
  }

  /*
   * 全局动画transform与popup自带的transform矛盾，要对动画参数进行细微订正
   */

  /* bottom => top */
  .effect() {
    animation-duration: 0.2s;
    animation-fill-mode: both;
    animation-timing-function: cubic-bezier(0.55, 0, 0.55, 0.2);
  }

  &-slide-up-enter,
  &-slide-up-appear {
    transform: translate(0, 0);
  }

  &-slide-up-enter,
  &-slide-up-appear,
  &-slide-up-exit {
    .effect();

    animation-play-state: paused;
  }

  &-slide-up-enter&-slide-up-enter-active,
  &-slide-up-appear&-slide-up-appear-active {
    animation-name: DoraPopupSlideUpIn;
    animation-play-state: running;
  }

  &-slide-up-exit&-slide-up-exit-active {
    animation-name: DoraPopupSlideUpOut;
    animation-play-state: running;
  }

  @keyframes DoraPopupSlideUpIn {
    0% {
      transform: translate(0, 0);
    }

    100% {
      transform: translate(0, -100%);
    }
  }

  @keyframes DoraPopupSlideUpOut {
    0% {
      transform: translate(0, -100%);
    }

    100% {
      transform: translate(0, 0);
    }
  }

  /* right => left */
  &-slide-left-enter,
  &-slide-left-appear {
    transform: translate(0, 0);
  }

  &-slide-left-enter,
  &-slide-left-appear,
  &-slide-left-exit {
    .effect();

    animation-play-state: paused;
  }

  &-slide-left-enter&-slide-left-enter-active,
  &-slide-left-appear&-slide-left-appear-active {
    animation-name: DoraPopupSlideLeftIn;
    animation-play-state: running;
  }

  &-slide-left-exit&-slide-left-exit-active {
    animation-name: DoraPopupSlideLeftOut;
    animation-play-state: running;
  }

  @keyframes DoraPopupSlideLeftIn {
    0% {
      transform: translate(0, 0);
    }

    100% {
      transform: translate(-100%, 0);
    }
  }

  @keyframes DoraPopupSlideLeftOut {
    0% {
      transform: translate(-100%, 0);
    }

    100% {
      transform: translate(0, 0);
    }
  }
}
