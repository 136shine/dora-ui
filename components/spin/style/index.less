@import '../../style/vars.less';

@spin-prefix-cls: dora-spin;
@spinner-prefix-cls: dora-spin-spinner;

.@{spin-prefix-cls} {
  position: relative;
  &-spinner-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background-color: @fill-mask;

    &__full {
      position: fixed;
      left: 0;
      top: 0;
      bottom: 0;
      right: 0;
      z-index: @toast-zindex;
    }
  }

  &-text {
    color: #fff;
    &-sm {
      padding-top: 4px;
      font-size: 12px;
      line-height: 12px;
    }

    &-md {
      padding-top: 6px;
      font-size: 14px;
      line-height: 14px;
    }

    &-lg {
      padding-top: 8px;
      font-size: 16px;
      line-height: 16px;
    }
  }
}
