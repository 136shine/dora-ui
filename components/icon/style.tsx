import '../styles/themes/default';
import '../styles/mixins';
import './style.less';
