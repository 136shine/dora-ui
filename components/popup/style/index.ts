import './index.less';
import '../../style/index';
