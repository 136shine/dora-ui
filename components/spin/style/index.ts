import './index.less';
import '../../icon/style';
import '../../style';
