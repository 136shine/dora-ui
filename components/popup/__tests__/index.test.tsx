import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { CSSTransition } from 'react-transition-group';
import Popup from '../index';
import '../style';

jest.mock('react-transition-group', () => {
  const FakeTransition = jest.fn(({ children }) => children);
  const FakeCSSTransition = jest.fn(props =>
    props.in ? (
      <FakeTransition>{props.children}</FakeTransition>
    ) : props.unmountOnExit ? null : (
      <div style={{ display: 'none' }}>{props.children}</div>
    )
  );
  return { CSSTransition: FakeCSSTransition, Transition: FakeTransition };
});

const App = ({ position = 'center' }) => {
  const [visible, setVisible] = React.useState(false);
  const [destroyOnClose, setDestroyOnClose] = React.useState(false);
  const [mask, setMask] = React.useState(true);

  return (
    <>
      <button onClick={() => setVisible(!visible)}>toggle modal visible</button>
      <button onClick={() => setDestroyOnClose(!destroyOnClose)}>
        toggle modal destroyOnClose
      </button>
      <button onClick={() => setMask(!mask)}>toggle modal mask</button>
      <Popup
        visible={visible}
        destroyOnClose={destroyOnClose}
        mask={mask}
        maskClosable
        onClose={() => setVisible(false)}
        position={position as any}
        wrapClassName="test-wrap-class"
      >
        content
      </Popup>
    </>
  );
};

describe('<Popup />', () => {
  test('should render outside container', () => {
    const { baseElement, getByText } = render(<App></App>);
    const visibleBtn = getByText(/visible/);

    expect(baseElement).toMatchSnapshot();

    fireEvent.click(visibleBtn);
    expect(baseElement).toMatchSnapshot();

    fireEvent.click(visibleBtn);
    expect(baseElement).toMatchSnapshot();
  });

  test('prop: visible ', () => {
    const { getByText } = render(<App></App>);
    const visibleBtn = getByText(/visible/);

    fireEvent.click(visibleBtn);
    expect(getByText(/content/)).toBeVisible();

    fireEvent.click(visibleBtn);
    expect(getByText(/content/)).not.toBeVisible();
  });

  test('prop: destroyOnClose', () => {
    const { getByText, queryByText } = render(<App></App>);
    const visibleBtn = getByText(/visible/);
    const destroyOnCloseBtn = getByText(/destroyOnClose/);

    fireEvent.click(visibleBtn);
    fireEvent.click(visibleBtn);
    expect(queryByText(/content/)).toBeInTheDocument();

    fireEvent.click(destroyOnCloseBtn);
    fireEvent.click(visibleBtn);
    fireEvent.click(visibleBtn);
    expect(queryByText(/content/)).not.toBeInTheDocument();
  });

  test('prop: mask', () => {
    const { getByText, baseElement } = render(<App></App>);
    const visibleBtn = getByText(/visible/);
    const maskBtn = getByText(/mask/);

    fireEvent.click(visibleBtn);
    expect(baseElement.getElementsByClassName('dora-popup__mask')).toHaveLength(1);
    fireEvent.click(visibleBtn);

    fireEvent.click(maskBtn);

    fireEvent.click(visibleBtn);
    expect(baseElement.getElementsByClassName('dora-popup__mask')).toHaveLength(0);
  });

  test('prop: maskCloseable onClose', () => {
    const { getByText, baseElement } = render(<App></App>);
    const visibleBtn = getByText(/visible/);

    fireEvent.click(visibleBtn);
    expect(getByText(/content/)).toBeVisible();

    fireEvent.click(baseElement.getElementsByClassName('dora-popup-mask')[0]);
    expect(getByText(/content/)).not.toBeVisible();
  });

  test('prop: position', () => {
    const positions = ['center', 'top', 'bottom', 'left', 'right'];

    positions.forEach(position => {
      const { baseElement, unmount } = render(<App position={position}></App>);
      const visibleBtn = screen.getByText(/visible/);

      fireEvent.click(visibleBtn);

      expect(baseElement.getElementsByClassName(`dora-popup__${position}`)).toHaveLength(1);

      unmount();
    });
  });

  test('prop: wrapClassName', () => {
    const { baseElement, getByText } = render(<App></App>);
    const visibleBtn = getByText(/visible/);

    fireEvent.click(visibleBtn);
    expect(baseElement.getElementsByClassName('dora-popup')[0]).toHaveClass('test-wrap-class');
  });
});
