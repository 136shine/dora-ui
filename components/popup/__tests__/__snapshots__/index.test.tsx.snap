// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> should render outside container 1`] = `
<body>
  <div>
    <button>
      toggle modal visible
    </button>
    <button>
      toggle modal destroyOnClose
    </button>
    <button>
      toggle modal mask
    </button>
  </div>
</body>
`;

exports[`<Popup /> should render outside container 2`] = `
<body
  style="position: fixed; top: 0px; left: 0px; right: 0px;"
>
  <div>
    <button>
      toggle modal visible
    </button>
    <button>
      toggle modal destroyOnClose
    </button>
    <button>
      toggle modal mask
    </button>
  </div>
  <div>
    <div
      class="test-wrap-class dora-popup dora-popup__center dora-popup__mask"
    >
      <div
        class="dora-popup-mask"
      />
      <div
        class="dora-popup-content"
      >
        content
      </div>
    </div>
  </div>
</body>
`;

exports[`<Popup /> should render outside container 3`] = `
<body
  style="top: 0px; left: 0px; right: 0px;"
>
  <div>
    <button>
      toggle modal visible
    </button>
    <button>
      toggle modal destroyOnClose
    </button>
    <button>
      toggle modal mask
    </button>
  </div>
  <div>
    <div
      class="test-wrap-class dora-popup dora-popup__center dora-popup__mask"
    >
      <div
        style="display: none;"
      >
        <div
          class="dora-popup-mask"
        />
      </div>
      <div
        style="display: none;"
      >
        <div
          class="dora-popup-content"
        >
          content
        </div>
      </div>
    </div>
  </div>
</body>
`;
